packages/server/src/unstable-core-do-not-import/router.ts

```diff
@@ -254,6 +282,30 @@ export function createCallerFactory<TRoot extends AnyRootTypes>() {
             getRawInput: async () => args[0],
             ctx,
             type: procedure._def.type,
+          }).catch(async (error) => {
+            if (error instanceof TRPCError) {
+              if (factoryOptions?.onError) {
+                await factoryOptions?.onError({
+                  ctx,
+                  error,
+                  input: args[0],
+                  path: fullPath,
+                  type: procedure._def.type,
+                });
+              }
+
+              if (options?.onError) {
+                await options?.onError({
+                  ctx,
+                  error,
+                  input: args[0],
+                  path: fullPath,
+                  type: procedure._def.type,
+                });
+              }
+            }
+
+            throw error;
```

Comment:
would an easier approach perhaps be to have a "default" onError handler that just rethrows the error, and if the user sets their own handler require them to rethrow the error if they choose to? i can see the case for the return type there though...

Reply:
It would save a few lines of code, but if the user can opt to not rethrow the error, any caller with an error handler set is no longer a transparent proxy for procedure calls from a pure perspective. (the return type would at the very least turn into T | undefined)
By always throwing, the pure part of the procedure call remains indifferent to the presence of an error handler, so if the error handler has no relevant side effects, it may just as well not exist from the perspective of the caller.
I suppose this could technically be enforced by changing the handler's return type into `never | Promise<never>`, but I don't think there's much of a benefit, and it would make the error handlers more verbose.